### DISCUSSION TITLE
I have a question concerning "Work has automated testing before merge to trunk"

### DISCUSSION THREAD

**Discussion by WoodyB** [Minimum-CD/cd-manifesto#377] [Q&A] [answered]
In the section describing the minimum activities required for CI we see this

The minimum activities required for CI are:

Trunk-based development
Work integrates to the trunk at a minimum daily
Work has automated testing before merge to trunk
Work is tested with other work automatically on merge
All feature work stops when the build is red
New work does not break delivered work

If short-lived branches are not being used and the team is directly pushing changes to the main trunk then I'm not clear about this activity, "Work has automated testing before merge to trunk"

My understanding of CI when not using short-lived branches is the small code change will be merged to the main trunk regardless of whether the  build and testing are successful or not.

Assuming my understand is correct about branchless CI does this just means the developers must always build and run locally before pushing a change to the main trunk and has nothing to do with the CI server not merging if the build or tests fail?

Plese help unconfuse me.

**Comment by bdfinst** [comment, marked as answer]
If we are doing CI correctly, we do not push untested code. The CI server needs the test to validate that nothing is broken, and without the tests, we have no way to protect production from bad code.

So, while someone may decide to:
- write code
- push it to a branch
- write tests, push those
- merge the branch

That workflow is catastrophic for mainline development.
